app/assets/javascripts/discourse/app/lib/codeblock-buttons.js

```diff
@@ -0,0 +1,193 @@
+        "click",
+        this._boundClickHandler
+      );
+
+      delete this._codeblockButtonClickHandlers[identifier];
+    }
+
+    this._codeblockButtonClickHandlers[identifier] = element;
+  }
+
+  _getCodeBlocks(element) {
+    let codeBlocks = [];
+    try {
+      codeBlocks = element.querySelectorAll(
+        ":scope > pre > code, :scope :not(article):not(blockquote) > pre > code"
+      );
+    } catch (e) {
+      // :scope is probably not supported by this browser
+      codeBlocks = [];
+    }
+    return codeBlocks;
+  }
+
+  _createButtons(codeBlocks) {
+    codeBlocks.forEach((codeBlock) => {
+      if (this.showFullscreen) {
+        const fullscreenButton = document.createElement("button");
+        fullscreenButton.classList.add("btn", "nohighlight", "fullscreen-cmd");
+        fullscreenButton.innerHTML = iconHTML("discourse-expand");
+        codeBlock.before(fullscreenButton);
+      }
+
+      if (this.showCopy) {
+        const copyButton = document.createElement("button");
+        copyButton.classList.add("btn", "nohighlight", "copy-cmd");
+        copyButton.innerHTML = iconHTML("copy");
+        codeBlock.before(copyButton);
+      }
+
+      codeBlock.parentElement.classList.add("codeblock-buttons");
+    });
+  }
+
+  _addClickEvent(element) {
+    element.addEventListener("click", this._boundClickHandler, false);
+  }
+
+  _handleClick(event) {
+    if (
+      !event.target.classList.contains("copy-cmd") &&
+      !event.target.classList.contains("fullscreen-cmd")
+    ) {
+      return;
+    }
+
+    const action = event.target.classList.contains("fullscreen-cmd")
+      ? "fullscreen"
+      : "copy";
+    const button = event.target;
+    const codeEl = button.parentElement.querySelector("code");
+
+    if (codeEl) {
+      // replace any weird whitespace characters with a proper '\u20' whitespace
+      const text = codeEl.innerText
+        .replace(
+          /[\f\v\u00a0\u1680\u2000-\u200a\u202f\u205f\u3000\ufeff]/g,
+          " "
+        )
+        .trim();
+
+      if (action === "copy") {
+        const result = clipboardCopy(text);
+        if (result.then) {
+          result.then(() => {
+            this._copyComplete(button);
+          });
+        } else if (result) {
+          this._copyComplete(button);
+        }
+      } else if (action === "fullscreen") {
+        showModal("fullscreen-code").setProperties({
+          code: text,
+          codeClasses: codeEl.className,
+        });
+      }
+    }
+  }
+
+  _copyComplete(button) {
+    button.classList.add("action-complete");
+    const state = button.innerHTML;
+    button.innerHTML = I18n.t("copy_codeblock.copied");
+
+    const commandId = guidFor(button);
+
+    if (this._fadeCopyCodeblocksRunners[commandId]) {
+      cancel(this._fadeCopyCodeblocksRunners[commandId]);
+      delete this._fadeCopyCodeblocksRunners[commandId];
+    }
+
+    this._fadeCopyCodeblocksRunners[commandId] = later(() => {
```

Comment:
I think you'll need the usual isDestroying/isDestroyed check here

Reply:
But this is a lib not a component?

Reply:
Ah, so no lifecycle props there. Carry on then.